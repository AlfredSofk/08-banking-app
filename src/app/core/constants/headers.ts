export const headers = {
    dispositivo: "PC",
    idioma: "es",
    uuid: "02e3eb27-6fb1-e542-e157-c301cc77ad2c",
    ip: "localhost",
    horaTransaccion: "string",
    llaveSimetrica: "xaqVyedHolrJB9vW4lIj5u9nuWIiaPpAQoOK4hm2j+Q=",
    vectorInicializacion: "Gz3MLPvKU1T5Pc3FfmNYPe9nuWIiaPpAQoOK4hm2j+Q="

}