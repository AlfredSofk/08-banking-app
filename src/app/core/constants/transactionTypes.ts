export enum TransactionTypes {
    WITHDRAW = 'withdraw',
    DEPOSIT = 'deposit',
    DEPOSIT_ACCOUNT = 'deposit-account',
    TRANSFER = 'transfer',
    PURCHASE_WEB = 'purchase-web',
    PURCHASE_STORE = 'purchase-store',
}