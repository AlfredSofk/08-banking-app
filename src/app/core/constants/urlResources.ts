import { environment } from "../../environment/environment";

export const urlResources = {
    generateToken: `${environment.apiBaseUrl}/utils/generate`,
    getClientData : `${environment.apiBaseUrl}/client/findByUserName`,
    getTransaction: (transactionType: string) => `${environment}/transactions/${transactionType}`,
    
};