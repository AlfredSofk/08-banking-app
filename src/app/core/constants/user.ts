export enum USER_LOGIN {
    username = "pablo",
    password = "12qwaszx"
}