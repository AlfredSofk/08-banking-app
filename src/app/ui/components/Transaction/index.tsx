
import { useForm } from 'react-hook-form';
import './style.scss'
import { IAppState } from '../../../core/interfaces/bankAccount';
import { TransactionTargets } from '../../../core/constants/transactionTypes';
import { Loader } from '../Loader/index';
import { useEffect } from 'react';


interface Props {
  state : IAppState
  titleTransaction: string
  transactionTarget: string
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  transactionHook: (data: any) => void
}


export const Transaction = ({ state,titleTransaction,  transactionTarget, transactionHook }: Props) => {

  const { register, handleSubmit, formState: { errors }, reset } = useForm();


  
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const onSubmit = (data: any) => {
    console.log(titleTransaction)
    console.log('Datos de retiro:', data);
    transactionHook(data)
    reset()
  };

  
  useEffect(() => {
    reset()
  }, [])

  return (
    <section className="withdrawal" aria-labelledby="withdrawal-heading">
      <h2 id="withdrawal-heading" className="withdrawal__title" data-testid="title-transaction">{titleTransaction} </h2>
      <form onSubmit={handleSubmit(onSubmit)} className="withdrawal__form" noValidate>

        <div className="withdrawal__form-group">
          <label htmlFor="accountNumber" className="withdrawal__label">{transactionTarget == TransactionTargets.TRANSFERENCIA ? 'Cuenta principal' : 'Número de Cuenta'}</label>
          <input
            id="accountNumber"
            type="text"
            className="withdrawal__input"
            {...register('accountNumber', {
              required: 'El número de cuenta es requerido',
              // pattern: {
              //   value: /^[0-9]}$/,
              //   message: 'El número de cuenta debe tener 10 dígitos'
              // }
            })}
            aria-invalid={errors.accountNumber ? "true" : "false"}
            aria-describedby="accountNumber-error"
          />
          {errors.accountNumber && (
            <span id="accountNumber-error" className="withdrawal__error" role="alert">
              {errors.accountNumber.message}
            </span>
          )}
        </div>
        
        {transactionTarget === TransactionTargets.TRANSFERENCIA  &&
          <div className="withdrawal__form-group">
            <label htmlFor="accountNumberReceiver" className="withdrawal__label">{transactionTarget == TransactionTargets.TRANSFERENCIA ? 'Cuenta destino' : 'Número de Cuenta'}</label>
            <input
              id="accountNumberReceiver"
              type="text"
              className="withdrawal__input"
              {...register('accountNumberReceiver', {
                required: 'El número de cuenta es requerido',
                // pattern: {
                //   value: /^[0-9]}$/,
                //   message: 'El número de cuenta debe tener 10 dígitos'
                // }
              })}
              aria-invalid={errors.accountNumberReceiver ? "true" : "false"}
              aria-describedby="accountNumberReceiver-error"
            />
            {errors.accountNumberReceiver && (
              <span id="accountNumberReceiver-error" className="withdrawal__error" role="alert">
                {errors.accountNumber.message}
              </span>
            )}
          </div>   
        }

        <div className="withdrawal__form-group">
          <label htmlFor="amount" className="withdrawal__label">{`Monto a ${transactionTarget}`}</label>
          <input
            id="amount"
            type="number"
            step="1"
            className="withdrawal__input"
            {...register('amount', {
              required: 'El monto es requerido',
              min: {
                value: 1,
                message: 'El monto debe ser mayor a 0',
              },
              valueAsNumber: true
            })}
            aria-invalid={errors.amount ? "true" : "false"}
            aria-describedby="amount-error"
          />
          {errors.amount && (
            <span id="amount-error" className="withdrawal__error" role="alert">
              {errors.amount.message}
            </span>
          )}
        </div>
          
          { state.loading 
            ? <Loader />
            :<button type="submit" className="withdrawal__submit" role='buttonTransaction'>{`Monto a ${transactionTarget}`}</button>
          }
      </form>
    </section>
  );
};
