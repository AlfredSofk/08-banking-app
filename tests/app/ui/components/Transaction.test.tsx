import { fireEvent, queryAllByRole, render, screen, waitFor } from '@testing-library/react';
import { Transaction } from '../../../../src/app/ui/components/Transaction/index';
import { vi } from 'vitest';
import { TransactionTargets } from '../../../../src/app/core/constants/transactionTypes';
import React from 'react';


describe('Test para el componente Transaction', () => {

    const mockTransactionHook = vi.fn();
    const mockState = {
        saldoGlobal: 10000,
        transacciones: [],
        username: 'testuser',
        userAccounts: [],
        loading: false,
        error: null,
    };

    afterEach(() => {
        vi.clearAllMocks();
    });

    test('Match Snapshot', () => {
        
        const {asFragment}  = render(<Transaction state={mockState} titleTransaction='test' transactionTarget={TransactionTargets.RETIRO} transactionHook={mockTransactionHook} />);
        expect(asFragment()).toMatchSnapshot();

    });

    test('Deberia mostrar los mensajes de error cuando se envía un formulario inválido de tipo retiro', async() => {

        const { getByRole, findAllByRole } = render( <Transaction state={mockState} titleTransaction='test' transactionTarget={TransactionTargets.RETIRO} transactionHook={mockTransactionHook} />);

        const buttonLogin = getByRole('buttonTransaction');

        fireEvent.submit(buttonLogin);

        expect(await findAllByRole('alert')).toHaveLength(2);
        expect(mockTransactionHook).not.toBeCalled()
    })

    test('Deberia mostrar los mensajes de error cuando se envía un formulario inválido de tipo deposito', async() => {

        const { getByRole, findAllByRole } = render( <Transaction state={mockState} titleTransaction='test' transactionTarget={TransactionTargets.DEPOSITO} transactionHook={mockTransactionHook} />);

        const buttonLogin = getByRole('buttonTransaction');

        fireEvent.submit(buttonLogin);

        expect(await findAllByRole('alert')).toHaveLength(2);
        expect(mockTransactionHook).not.toBeCalled()
    })

    test('Deberia mostrar los mensajes de error cuando se envía un formulario inválido de tipo deposito', async() => {

        const { getByRole, findAllByRole } = render( <Transaction state={mockState} titleTransaction='test' transactionTarget={TransactionTargets.COMPRA} transactionHook={mockTransactionHook} />);

        const buttonLogin = getByRole('buttonTransaction');

        fireEvent.submit(buttonLogin);

        expect(await findAllByRole('alert')).toHaveLength(2);
        expect(mockTransactionHook).not.toBeCalled()
    })

    test('Deberia mostrar los mensajes de error cuando se envía un formulario inválido de tipo deposito', async() => {

        const { getByRole, findAllByRole } = render( <Transaction state={mockState} titleTransaction='test' transactionTarget={TransactionTargets.TRANSFERENCIA} transactionHook={mockTransactionHook} />);

        const buttonLogin = getByRole('buttonTransaction');

        fireEvent.submit(buttonLogin);

        expect(await findAllByRole('alert')).toHaveLength(3);
        expect(mockTransactionHook).not.toBeCalled()
    })

    test('No deberia mostrar un mensaje de error cuando los datos son validos retiro', async() => {        
        const { getByRole, getByLabelText, queryAllByRole } = render( <Transaction state={mockState} titleTransaction='retiro ATM' transactionTarget={TransactionTargets.RETIRO} transactionHook={mockTransactionHook} />);

        const accountNumberMock : string = "123456789"
        const amountMock : number = 100

        await waitFor(() =>  {
              fireEvent.input(getByRole("textbox", { name: /número de cuenta/i }), {
                  target: {
                    value: accountNumberMock,
                  },
                })
      
                fireEvent.input(getByLabelText("Monto a retirar"), {
                  target: {
                    value: amountMock,
                  },
                })
      
                fireEvent.submit(getByRole('buttonTransaction'));

            expect(queryAllByRole("alert")).toHaveLength(0)
            expect(mockTransactionHook).toBeCalledWith({ accountNumber: accountNumberMock, amount: amountMock })
        })
        expect (getByRole("textbox", { name: /número de cuenta/i })).toHaveValue(accountNumberMock)
    });    

    test('No deberia mostrar error cuando los datos de la trasferencia son validos', async() => {        
        const { getByRole, getByLabelText, queryAllByRole } = render( <Transaction state={mockState} titleTransaction='transferencia' transactionTarget={TransactionTargets.TRANSFERENCIA} transactionHook={mockTransactionHook} />);
        const accountNumberMock : string = "123456789"
        const amountMock : number = 100
        const accountNumberReceiverMock : string = "987654321"
        await waitFor(() =>  {
              fireEvent.input(getByRole("textbox", { name: /cuenta principal/i }), {
                  target: {
                    value: accountNumberMock,
                  },
                })
      
                fireEvent.input(getByLabelText("Monto a transferir"), {
                  target: {
                    value: amountMock,
                  },
                })
      
                fireEvent.input(getByRole("textbox", { name: /cuenta destino/i }), {
                  target: {
                    value: accountNumberReceiverMock,
                  },
                })
      
                fireEvent.submit(getByRole('buttonTransaction'));
      
            expect(queryAllByRole("alert")).toHaveLength(0)
            expect(mockTransactionHook).toBeCalledWith({ accountNumber: accountNumberMock, amount: amountMock, accountNumberReceiver: accountNumberReceiverMock })
        })
        expect (getByRole("textbox", { name: /cuenta principal/i })).toHaveValue(accountNumberMock)
        expect (getByRole("textbox", { name: /cuenta destino/i })).toHaveValue(accountNumberReceiverMock)
    });
});                     
