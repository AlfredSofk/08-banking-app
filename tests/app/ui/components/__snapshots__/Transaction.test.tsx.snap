// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Test para el componente Transaction > Match Snapshot 1`] = `
<DocumentFragment>
  <section
    aria-labelledby="withdrawal-heading"
    class="withdrawal"
  >
    <h2
      class="withdrawal__title"
      data-testid="title-transaction"
      id="withdrawal-heading"
    >
      test 
    </h2>
    <form
      class="withdrawal__form"
      novalidate=""
    >
      <div
        class="withdrawal__form-group"
      >
        <label
          class="withdrawal__label"
          for="accountNumber"
        >
          NÃºmero de Cuenta
        </label>
        <input
          aria-describedby="accountNumber-error"
          aria-invalid="false"
          class="withdrawal__input"
          id="accountNumber"
          name="accountNumber"
          type="text"
        />
      </div>
      <div
        class="withdrawal__form-group"
      >
        <label
          class="withdrawal__label"
          for="amount"
        >
          Monto a retirar
        </label>
        <input
          aria-describedby="amount-error"
          aria-invalid="false"
          class="withdrawal__input"
          id="amount"
          name="amount"
          step="1"
          type="number"
        />
      </div>
      <button
        class="withdrawal__submit"
        role="buttonTransaction"
        type="submit"
      >
        Monto a retirar
      </button>
    </form>
  </section>
</DocumentFragment>
`;
